Fix for scrapinghub#116

adityas114 opened the pull request:
Fixes #116. I fixed the order in rdfa.py by checking the correct order in the document object. I also fixed test 'test_rdfa_not_preserving_order'.

ivanprado commented:
Thank you @adityas114 for your contribution! I'll have a look in the following days. I already have a quick look and have some comments:
* I wonder if the approach is the correct. I mean, the proposed code is patching somehow the result from the source library (RDFlib or pyRDFa, not sure. See https://github.com/scrapinghub/extruct/issues/116#issuecomment-500171872). A better fix would be to fix the original library. But also seems that extucts pinned the rdflib version probably because newer versions introduces some incompatibilities. So then maybe patching is the right thing to do. Not sure. I'll see the code during the next days and probably I'll have it clearer.
* I see that the code is iterating several times over the tree. I wonder how this is impacting the performance. Could you measure the impact on the performance of your changes? Also, wonder if speed could be improved if you use xpath expressions in some places instead of transversing the tree. 
* There are some parts of your code that remain untested. See the codecov report: https://codecov.io/gh/scrapinghub/extruct/pull/139/diff?src=pr&el=tree#diff-ZXh0cnVjdC9yZGZhLnB5. Tests covering them will be required.

lopuhin commented:
> But also seems that extucts pinned the rdflib version probably because newer versions introduces some incompatibilities. So then maybe patching is the right thing to do. 

https://github.com/scrapinghub/extruct/issues/135 is also related

adityas114 commented:
Thank you @ivanprado for you comments. 

* I agree that ideally it should be fixed in the original library, but I believe that maybe we can patch it in extruct itself for the time being.
* Yes xpath may improve performance, but I am not familiar with it so I didn't use it for now. I'll look into it.
* Okay I'll add more tests where required.

Let me know if there is any more feedback 😄

adityas114 commented:
@ivanprado 

I have added tests to improve code coverage, and I have used xpath to improve efficiency. 

For testing, I edited 'elysianfields.html', however after changing two other tests fail. From what I understand, it seems that there is an issue with the implementation of open graph protocol. I have added xfail tags for those 2 tests. 

Let me know what you think 😄

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+        # if namespace taken from prefix attribute in head tag
+        if match:
+            ns = match.group().split(': ')[1]
+            return ns + prop.split(':')[1]
```
If prop is wrong (e.g. "baddefinedprop") this line would fail (index out of range).

adityas114 replied:
The if condition will only pass if the property is well defined - in this case only if the namespace was in the prefix attribute of the head tag. I tried with a wrong prop ("baddefinedprop"), and it worked fine.

ivanprado replied:
I see, good point. I missed that.

ivanprado replied:
Oh right, I wouldn't fail for "baddefinedprop". But it would fail for "owl" property if head element is something like:
```html
<head prefix="owl:http://www.w3.org/2002/07/owl#">
```

ivanprado replied:
I think this is now solved.

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+            
+            for key in keys:
+                if type(json_object[key]) is list and len(json_object[key]) > 1:
+                    ordered = list(map(lambda meta_tag: meta_tag.get('content'), head_element.xpath("meta[@property='" + key + "']")))
```
The line is too long, consider splitting it. Same for other very long lines.

adityas114 replied:
Okay, I'll do that.

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+        for json_object in json_objects:
+            keys = json_object.keys()
+
+            ordered = []
```
This is not required

adityas114 replied:
In all my tests, it always returned a list of size 1, with the first element in the list as the desired json. I was not sure about this though - will it always return a list of size 1?

ivanprado replied:
I don't get your point here :thinking: What I mean is that the variable `ordered` is initialized here as an empty list, but then this value is overwritten at line:
https://github.com/adityas114/extruct/blob/23425a0139db5973fb0d3af926d433c8375c393b/extruct/rdfa.py#L124
So it seems this initialized value is not used anywhere in any case.

adityas114 replied:
Sorry, I thought you were talking about the for loop. Okay I see the problem, I am more used to programming in Java 😄. I'll fix it.

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+                    ordered = list(map(lambda meta_tag: meta_tag.get('content'), head_element.xpath("meta[@property='" + key + "']")))
+
+                    self.sort(json_object, ordered, key)
+                    ordered.clear()
```
This is not required.

ivanprado replied:
Same than in the last comment. Here you're clearing the ordered list... but this cleared list will be completely overwritten at line https://github.com/adityas114/extruct/blob/23425a0139db5973fb0d3af926d433c8375c393b/extruct/rdfa.py#L124. 
So it seems the line can be completely removed.

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+        return prop
+    
+    # sorts the rdfa tags in jsonld string
+    def sort(self, json_object, ordered, key):
```
Consider changing this sort algorithm by something like: 
```py
    def sort(self, json_object, ordered, key):
        idx_for_value = dict((value, idx) for idx, value in enumerate(ordered))
        json_object[key].sort(key=lambda props: idx_for_value.get(props.get('@value'), len(ordered)))
```

Also probably better to receive `json_object[key]` directly to save one function parameter

adityas114 replied:
Okay I'll optimise this.

ivanprado commented on tests/test_extruct.py:
```diff
@@ -13,26 +13,32 @@ class TestGeneric(unittest.TestCase):
 
     maxDiff = None
 
+    @pytest.mark.xfail
```
Why is this marked as xfail?

adityas114 replied:
I made some changes in the "tests/samples/songkick/elysianfields.html" file. I added a few corner cases to test my code properly, but this test was failing. I inspected, and the functionality for the open graph protocol is not correct. You may want to test this yourself.

ivanprado replied:
I checked. I see that `test1` images were not in the output, and it is because `test1: http://example.org/test1#` is not in the prefix definition. But it is in `xmlns` definition, so it might work as well, so maybe you found an issue here. I visit https://ogp.me/ and examples there uses only the `prefix` but as they are based in RDFa I guess it should also support `xmlns`

But if such is the case, better to treat it as an independent issue to be fixed in a different PR in the future. I propose that you add  `test1: http://example.org/test1#` to the prefixes so that the test passes.

adityas114 replied:
I added it to test all my corner cases. I will then need to add another sample file, and I was not sure if that is ideal. If you want me to do that, should I add a new file in the songkick folder itself?

ivanprado replied:
Yes, that sounds good. Maybe better to have another example to test your corner cases, and leave this file Ok for the global checks.

In any case with your fix we should be able to remove these lines safely also:
https://github.com/adityas114/extruct/blob/ab3d34e5b5eeed9611e49ee06cc6816d9692988c/tests/test_extruct.py#L21-L29

ivanprado commented on tests/test_uniform.py:
```diff
@@ -9,6 +11,7 @@ class TestUniform(unittest.TestCase):
 
     maxDiff = None
 
+    @pytest.mark.xfail
```
Why is this marked as xfail now?

adityas114 replied:
Same as before - the code for the opengraph protocol is not correct.

ivanprado replied:
I propose the same here than I proposed here: https://github.com/scrapinghub/extruct/pull/139/files#r440913833

ivanprado commented:
Something also remaining:

- Documenting clearly that this is a hack to fix the ordering and that it should be disabled once PyRDFA fixes itself.
- Adding more tests with corner cases (e.g. bad namespace)

adityas114 commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+
+        # if namespace taken from xmlns attribute in html tag
+        if ('xmlns:' + prefix) in html_element.keys():
+            return html_element.get('xmlns:' + prefix) + prop.split(':')[1]
```
Like before, the if condition won't pass if the prop is not well defined.

ivanprado replied:
I think it can happen. For property "owl" and HTML element:
```html
<HTML xmlns:owl="http://www.w3.org/2002/07/owl#">
```

adityas114 replied:
Okay, I'll fix for this case.

adityas114 commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,104 @@
+
+        # if namespace present in inital context
+        if prefix in context:
+            return context[prefix] + prop.split(':')[1]
```
Like before, the if condition won't pass if the prop is not well defined.

ivanprado replied:
It will fail if property is "owl" for example.

ivanprado commented:
Hi @adityas114 . 

Firstly, thank you again for collaborating with this fix. It is very appreciated :+1: 

I left some additional comments.

adityas114 commented:
Hi @ivanprado .

I added some changes, please check them out. I believe I have incorporated all your suggestions - let me know if I missed something 😄 

Also, I also did a test to compare the performance. I called the extract method 100 times, with the 'elysianfields_1.html' file as parameter. Without patch, it ran in 13.070286s, and with the patch it ran in 13.093408s. Clearly, the performance is only marginally affected.

Let me know if you have more suggestions, I'd be happy to help 👍 😄

ivanprado commented on tests/test_extruct.py:
```diff
@@ -17,22 +17,18 @@ def test_all(self):
         self.assertEqual(jsonize_dict(data), expected)
 
-    @pytest.mark.xfail
     def test_rdfa_not_preserving_order(self):
```
Could you rename this method to something like `test_rdfa_is_preserving_order`?

ivanprado commented on tests/test_extruct.py:
```diff
@@ -17,22 +17,18 @@ def test_all(self):
     def test_rdfa_not_preserving_order(self):
         # See https://github.com/scrapinghub/extruct/issues/116
         # RDFa is not preserving ordering on duplicated properties. So this
         # test sometimes fails for property 'http://ogp.me/ns#image'
```
Could you remove this part of the comment? It should never fail now.

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,100 @@
+    
+    # fixes order of rdfa tags in jsonld string
+    # by comparing with order in document object
+    def fixOrder(self, jsonld_string, document):
```
Could you rename it to `fix_order` to follow the convention?

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,100 @@
+    
+    # sorts the rdfa tags in jsonld string
+    def sort(self, unordered, ordered):
+        idx_for_value = dict((value, idx) for idx, value in enumerate(ordered))
```
Can you change that by 
```suggestion
        idx_for_value = dict(reversed((value, idx) for idx, value in enumerate(ordered)))
```

And include a case with repeated values like:
```html
    <meta property="test1:image" content="http://images.sk-static.com/images/media/img/col4/20100330-103600-169450.jpg">
    <meta property="test1:image" content="http://images.sk-static.com/SECONDARY_IMAGE.jpg">
    <meta property="test1:image" content="http://images.sk-static.com/images/media/img/col4/20100330-103600-169450.jpg">
```

adityas114 replied:
Why do we need this change? 

Also, for repeated values I see another problem - only one value is returned by PyRDFA. How should we go about this problem? This seems to be another issue altogether?

ivanprado replied:
> only one value is returned by PyRDFA

That's is not so problematic. If the content is repeated (e.g. the same image URL), then returning just the first one is OK. 

> Why do we need this change?

To ensure that the order in the presence of duplicates is from the first seen result. In the current code, the opposite is happening.

adityas114 replied:
> To ensure that the order in the presence of duplicates is from the first seen result. In the current code, the opposite is happening.

Can you verify if it works correctly for you? For me it throws `TypeError: 'generator' object is not reversible`

ivanprado replied:
Try with:
`idx_for_value = dict(reversed([(value, idx) for idx, value in enumerate(ordered)]))`

adityas114 replied:
It works now 👍

ivanprado commented on extruct/rdfa.py:
```diff
@@ -28,6 +29,100 @@
+        head_element = document.xpath('//head')[0]
+                
+        for meta_tag in head_element.xpath("meta[@property]"):
+            meta_tag.attrib['property'] = self.replaceNS(meta_tag.attrib['property'], 
```
I see now a problem here. This is actually modifying the `document` itself, and can have then side effects in other parts of the code. I think it can be avoided by just building a `defaultdict` with the properties and its content, and use it below in the the code. Could you please change it?

adityas114 replied:
I see the problem now. I am not sure how to use `defaultdict` though. Will making a deep copy of the `document` object be sufficient?

ivanprado replied:
Making a copy of the document is very inefficient. You don't need a `defaultdict`, you can use a simple `dict` for it but `defaultdict` makes your life easier: https://www.accelebrate.com/blog/using-defaultdict-python#:~:text=Dictionaries%20are%20a%20convenient%20way,retrieval%20by%20name%20(key).&text=A%20defaultdict%20works%20exactly%20like,will%20never%20raise%20a%20KeyError.

ivanprado commented on tests/samples/songkick/elysianfields.html:
```diff
@@ -585,7 +585,7 @@ <h2>Posters (8)</h2>
 
           <div class="component events-summary" id="gigography-summary">
-  <h2 class="calendar">Past concerts (321) <small><a rel="nofollow" href="/artists/236156-elysian-fields/gigography">See all</a></small></h2>
+  <h2 class="calendar">Past concerts (321) <small><a rel="nofollow" href="/artists/236156-elysian-fields/gigography">See all</a></small></h2>
```
This is reporting like a strange change in a space. Maybe a encoding issue. Could you just revert this file to the master state so that this space is as it was?.

adityas114 replied:
I am not very comfortable with git and not sure how to do this, can you help me with this? I am working with the GitHub GUI.

I originally made my changes to `elysianfields.html`, then renamed it to `elysianfields_1.html`, and created a new file `elysianfields.html` by copying the code from the master branch.

ivanprado replied:
I tried to do in this commit https://github.com/scrapinghub/extruct/pull/139/commits/66c7b89e57ae2f3c134b5dd90e20ad6f7def45f9 but Github still reports a difference. Anyway, this is not important, so let's forget about it :+1:

ivanprado commented on tests/test_uniform.py:
```diff
@@ -1,5 +1,7 @@
 import unittest
 
+import pytest
```
Is is import required? If not, could you please remove it?

adityas114 commented:
I have made all the suggested changes. Please review 😄

lopuhin commented on extruct/rdfa.py:
```diff
@@ -16,6 +17,8 @@
 
 from extruct.utils import parse_xmldom_html
 
+from collections import defaultdict
```
Since it's an stdlib import, it's best to place it at the top of the file

lopuhin commented on extruct/rdfa.py:
```diff
@@ -51,4 +153,10 @@ def extract_items(self, document, base_url=None, expanded=True):
+            # hack to fix the ordering of multi-value properties (see issue 116)
+            # it should be disabled once PyRDFA fixes itself
+            return self._fix_order(jsonld_string, document)
+        except:
```
Why all exceptions are ignored here? Is it possible to be more explicit with errors?

ivanprado replied:
I'll answer that one. The idea here is that `_fix_order` is just a temporal hack to be removed once `rdflib` fixes the ordering issue somehow. So the idea was that this code should never harm original implementation. Here is where it is ensured: In case any exception is raised by `_fix_order`, the original implementation should be invoked as a fallback. There are also some unit tests that check for that.

lopuhin replied:
I see, thanks for clarification @ivanprado . Good that we also have tests which make sure that the re-ordering works.

lopuhin commented on extruct/rdfa.py:
```diff
@@ -35,6 +38,105 @@
 
 class RDFaExtractor(object):
 
+    def _replaceNS(self, prop, html_element, head_element):
```
I'm curious if there are any helpers in ``rdflib`` which could be used instead?

adityas114 replied:
I looked for it, but I couldn't find anything. I was able to find the initial context, which I have used already.

lopuhin replied:
I see, thanks for checking 👍
